lib/stack/src/layer.rs

```diff
@@ -10,15 +10,15 @@ use std::marker::PhantomData;
 /// ```ignore
 /// impl<M: Stack<SocketAddr>> Layer<Authority, SocketAddr, M> for BalanceLayer<M> { ... }
 /// ```
-pub trait Layer<T, U, M: super::Stack<U>> {
+pub trait Layer<T, U, S: super::Stack<U>> {
     type Value;
     type Error;
     type Stack: super::Stack<T, Value = Self::Value, Error = Self::Error>;
 
-    /// Produce a `Stack` value from a `M` value.
-    fn bind(&self, next: M) -> Self::Stack;
+    /// Produces a `Stack` value from a `M` value.
+    fn bind(&self, next: S) -> Self::Stack;
 
-    /// Compose this `Layer` with another.
+    /// Produces a new Layer with this layer wrapping the provided inner layer.
```

Comment:
This is minor, but I think it would be good to have a doc comment here explaining the cases where `and_then` can make compile times slow/when you should use `push` instead?

Reply:
I honestly don't understand it well enough to have a cogent explanation.

Reply:
Okay, that's fair. I think even a "This sometimes makes type inference slow, use `push` instead" would be sufficient, though...